### DISCUSSION TITLE
Çakıl taşlari ve metal kutular

### DISCUSSION THREAD

**Discussion by sonsuzus** [sonsuzus/sonsuzus#453] [Soru Cevap]
Her birinin içinde farklı sayıda çakıl taşı bulunan soldan sağa doğru bir çizgide yere sabitlenmiş N tane metal kutumuz olduğunu düşünelim. Herhangi iki kutunun içinde aynı sayıda çakıl taşı bulunmamaktadır ve metal kutular hareket ettirilememektedir. Amacımız çakıl taşlarının kutulara dağılım sayılarını değiştirmeden, soldan sağa gittiğimizde kutulardaki çakıl taşları sayıları küçükten büyüğe artacak şekilde çakıl taşlarının kutularda bulunmasını sağlamaktır. Elimizde kağıt kalem olmadığından ve çakıl taşlarını kutular dışında bir yerde bekletirsek oluşabilecek karışıklığı engellemek için aşağıdaki gibi bir yöntem (algoritma) kullanmaya karar veriyoruz.

1. Soldan sağa giderek ardışık bütün kutu çiftlerine bak. Yani sırasıyla 1. ve 2. kutu, 2. ve 3. kutu, ...., N-1. ve N. kutuya bak. Bu sırada bir çift kutudan bir sonraki çifte geçmeden önce, eğer bu kutu çiftinde soldaki kutu sağdaki kutudan daha fazla çakıl taşı içeriyorsa bu iki kutudaki çakıl taşı sayıları arasındaki fark kadar çakıl taşını soldaki kutudan sağdaki kutuya aktar.
2. 1 adımda eğer herhangi iki kutu arasında çakıl taşı aktarması yaptı isen 1. adıma geri dön ve en baştan kutuların sonuna kadar aynı adımı tekrar et. Eğer 1. adımda bir çakıl taşı aktarması yapmadı isen zaten artık kutulardaki çakıl taşları soldan sağa küçükten büyüğe
yerleşmiş demektir.

**Comment by sonsuzus** [comment]
#### Soru 1

Eğer başlangıçta 7 tane metal kutuda soldan sağa 9, 4, 5, 3, 6, 1, 2 tane çakıl taşı varsa algoritmamızın birinci turu sonunda yani kutuları soldan sağa bir kez geçtiğimizde toplam kaç tane kutu ikilisi arasında çakıl taşı değişimi yapmış oluruz?

**Reply by sonsuzus** [reply]
9 baştan başlayıp son kutuya gidene kadar değerini koruyacağı için toplam 6 defa değişim yapmış oluruz.

**Comment by sonsuzus** [comment]
#### Soru 2

Eğer başlangıçta 7 tane metal kutuda soldan sağa 9, 4, 5, 3, 6, 1, 2 tane çakıl taşı varsa algoritmamızın birinci turu sonunda yani kutuları soldan sağa bir kez geçtiğimizde toplam kaç tane çakıl taşını bir kutudan diğerine taşımış oluruz? (Not: aynı çakıl taşı birden fazla kez taşınabilir, bu soruda her taşımayı ayrı ayrı saymamız isteniyor.)

**Reply by sonsuzus** [reply]
Aradaki farkların toplamı kadar çakıl taşı taşınacaktır. Yani ilk değişimde 5 çakıl taşı, ikincide 4, sonrasında 6,, 3, 8 ve 7 olmak üzere toplamda 5+4+6+3+8+7 = 33 çakıl taşı yer değiştirecektir.

**Comment by sonsuzus** [comment]
#### Soru 3

Eğer başlangıçta 7 tane metal kutuda soldan sağa 9, 4, 5, 3, 6, 1, 2 tane çakıl taşı varsa toplam kaç turda çakıl taşlarını küçükten büyüğe yerleştirme işini bitirmiş oluruz? (Hiç çakıl taşı değişimi yapılmayan son kontrol turunu toplam tur sayısına dahil etmeyi unutmayınız)

**Reply by Burak2210** [reply]
1. tur: 4 5 3 6 1 2 9
2. tur: 4 3 5 1 2 6 9
3. tur: 3 4 1 2 5 6 9
4. tur: 3 1 2 4 5 6 9
5. tur: 1 2 3 4 5 6 9
6. tur: 1 2 3 4 5 6 9 (kontrol)

**Comment by sonsuzus** [comment]
#### Soru 4

Eğer başlangıçta 8 tane metal kutuda soldan sağa 7, 5, 8, 9, 4, 3, 6, 2 tane çakıl taşı varsa algoritmamızın birinci turu sonunda yani kutuları soldan sağa bir kez geçtiğimizde toplam kaç tane kutu ikilisi arasında çakıl taşı değişimi yapmış oluruz?

**Reply by Burak2210** [reply]
5 kez. 5 7 8 4 3 6 2 9
İlk olarak 5 ve 7 değişiyor. Sonrasında 9 sayısı 4 kez değişiyor.

**Comment by sonsuzus** [comment]
#### Soru 5

Eğer başlangıçta 8 tane metal kutuda soldan sağa 7, 5, 8, 9, 4, 3, 6, 2 tane çakıl taşı varsa toplam kaç turda çakıl taşlarını küçükten büyüğe yerleştirme işini bitirmiş oluruz? (Hiç çakıl taşı değişimi yapılmayan son kontrol turunu toplam tur sayısına dahil etmeyi unutmayınız)

**Reply by Burak2210** [reply]
1. Tur: 5 7 8 4 3 6 2 9
2. Tur: 5 7 4 3 6 2 8 9
3. Tur: 5 4 3 6 2 7 8 9
4. Tur: 4 3 5 2 6 7 8 9
5. Tur: 3 4 2 5 6 7 8 9
6. Tur: 3 2 4 5 6 7 8 9
7. Tur: 2 3 4 5 6 7 8 9
8. Tur: 2 3 4 5 6 7 8 9 (kontrol)
